File: Data Structures & Algorithms/trapping-rain-water/submission-4.py
class Solution:
    def trap(self, heights: List[int]) -> int:
        n = len(heights)
        left = 1
        right = n - 2
        lmax = heights[left-1]
        rmax = heights[right+1]
        res = 0
        while(left<=right):
            if rmax<=lmax:
                res += max(0,rmax - heights[right])
                rmax = max(rmax, heights[right])
                right -= 1
            else:
                res += max(0, lmax-heights[left])
                lmax = max(lmax, heights[left])
                left += 1
        return res

File: Data Structures & Algorithms/is-palindrome/submission-4.cpp
class Solution {
public:
    bool isPalindrome(string s) {
        
        string t = "";
        for(char c : s) {
            if(isalnum(c)) {
                t += tolower(c);
            }
        }
        if(t.length() == 1 || t.length() == 0) return true;
        int len = t.size();
        int left = 0;
        int right = len-1;
        bool ans = false;
        while(left < right) {
            if(t[left] == t[right]) {
                ans = true;
                left++;
                right--;
            }
            else {
                return false;
            }
        }
        return ans;
    }
};
